pytorch_lightning/trainer/trainer.py

```diff
@@ -1171,15 +1174,15 @@ def _run(
             self._call_callback_hooks("on_fit_end")
             self._call_lightning_module_hook("on_fit_end")
 
-        # teardown if necessary (similar calls for spawn plugins are excluded as they have
-        # been included at the end of `new_process` functions)
-        if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
-            self._call_teardown_hook()
+        self._call_teardown_hook()
 
         if self.state.status != TrainerStatus.INTERRUPTED:
             self.state.status = TrainerStatus.FINISHED
         self.state.stage = None
 
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+            results = self.training_type_plugin._collect_rank_zero_results(self, results)
```

Comment:
The _collect_rank_zero_results called after teardown hook, will there be issue if processes got killed in teardown hook?

Reply:
_collect_rank_zero_results needs to be at the very and, because it needs to include the latest state of trainer.state. Whether teardown runs before or after this does not matter at the moment (process killing does not happen inside teardown). We may want to collect ideas how to polish this in #10985

Reply:
But if the processes already teardown, will we lost state_dicts?
The old new_process logic seems collected the output first? I'm probably miss something here
https://github.com/PyTorchLightning/pytorch-lightning/blob/aeb0b5595fd73d086f4ae0f99d3f1f112f6a4c29/pytorch_lightning/plugins/training_type/ddp_spawn.py#L202-L205

Reply:
The teardown() call does not kill the process. Teardown() is just calling accelerator.teardown() etc. and teardown() hooks on LM etc. runs there. 

Here is what happens inside _run():

```py
def _run(...):
     # each worker process runs this function
     # <--- worker process starts here

     setup()
     ...
     run_stage()
     ...
     teardown() 

     _collect_rank_zero_results()

     # end of the function, all worker processes are done and will shut down
     return
```